crates/ruff_python_formatter/src/expression/string/docstring.rs

```diff
@@ -702,19 +784,336 @@ impl<'src> CodeExampleDoctest<'src> {
+        let (opening_indent, rest) = indent_with_suffix(original.line);
+        if rest.starts_with(".. ") {
+            if let Some(litblock) = CodeExampleRst::new_code_block(original) {
+                return Some(litblock);
+            }
+            // In theory, we could still have something that looks like a literal block,
+            // but if the line starts with `.. `, then it seems like it probably shouldn't
+            // be a literal block. For example:
+            //
+            //     .. code-block::
+            //
+            //         cool_stuff( 1 )
+            //
+            // The above is not valid because the `language` argument is missing from
+            // the `code-block` directive. Because of how we handle it here, the above
+            // is not treated as a code snippet.
+            //
+            // ... but it
+            return None;
+        }
+        // At this point, we know we didn't find a code block, so the only
+        // thing we can hope for is a literal block which must end with a `::`.
+        if !rest.trim_end().ends_with("::") {
+            return None;
+        }
+        Some(CodeExampleRst {
+            lines: vec![],
+            opening_indent: indentation_length(opening_indent),
+            min_indent: None,
+            is_directive: false,
+        })
+    }
+
+    /// Attempts to create a new reStructuredText code example from a
+    /// `code-block` or `sourcecode` directive. If one couldn't be found, then
+    /// `None` is returned.
+    fn new_code_block(original: InputDocstringLine<'src>) -> Option<CodeExampleRst> {
+        // This regex attempts to parse the start of a reStructuredText code
+        // block [directive]. From the reStructuredText spec:
+        //
+        // > Directives are indicated by an explicit markup start (".. ")
+        // > followed by the directive type, two colons, and whitespace
+        // > (together called the "directive marker"). Directive types
+        // > are case-insensitive single words (alphanumerics plus
+        // > isolated internal hyphens, underscores, plus signs, colons,
+        // > and periods; no whitespace).
+        //
+        // The language names matched here (e.g., `python` or `py`) are taken
+        // from the [Pygments lexer names], which is referenced from the docs
+        // for the [code-block] directive.
+        //
+        // [directives]: https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#directives
+        // [Pygments lexer names]: https://pygments.org/docs/lexers/
+        // [code-block]: https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-code-block
+        static DIRECTIVE_START: Lazy<Regex> = Lazy::new(|| {
+            Regex::new(
+                r"(?m)^\s*\.\. \s*(?i:code-block|sourcecode)::\s*(?i:python|py|python3|py3)$",
+            )
+            .unwrap()
+        });
```

Comment:
If I remember correctly, the Regex multithreading performance cost has been mitigated, right? 

We could otherwise use our `Cursor` implementation for parsing out the string. Although skipping the whitespace could be somewhat annoying

Reply:
Mostly mitigated, yeah. And this only runs on lines that start with `.. `. So it should be very infrequently called.